docs/DesignDocs/system_ocall_opt_in.md

```diff
@@ -0,0 +1,120 @@
+# Explicit Enclave Opt-in to System OCalls
+
+System OCalls are OCalls that are part of Open Enclave core libraries. These are generally not
+called by the enclave developer but instead are called by other functions internal to OE SDK.
+
+Because OCalls have potential security impact, an enclave developer should be able to opt-out
+of these OCalls being made and should have visibility into what OCalls might be called by OE
+SDK.
+
+## EDL installation
+
+In order to support ocall opt-in, system EDL files will now be installed as part of the SDK at
+`${install_path}/include/openenclave/edl/`. This allows them to be imported into user edl by the
+enclave developer.
+
+## Developer opt-in
+
+In order to use any system OCall, it must be imported into the enclave's EDL file. If it is not
+included, but the OCall is still called by the enclave application or some part of the SDK,
+enclave linking will fail.
```

Comment:
Why linking rather than making compiling fail?  (When we did this in the new_platforms branch last year, I believe it was the compile step that would fail, because the prototype for the called API would not be found)

Reply:
Oh you're right, compiling would fail because the function wouldn't be present in the generated header.

I will update the doc.

Reply:
Actually let me take that back.

If the developer doesn't explicitly call any of these functions (as is the intention) then compilation will succeed just fine. The functions are just called by other OE libraries (such as oelibc). So in this example linking would fail when liboelibc.a is linked into the enclave binary if the proper edl file isn't imported.

Reply:
Got it, thanks

Reply:
Ok so the core issue then is that libc is monolithic (it wasn't in our new_platforms branch which is why we got compiler errors instead of linker errors).  So if you don't use any sockets APIs, it sounds like right now all the sockets apis are still in libc, making your TCB potentially larger with code that is never called by the enclave app, instead of having more granular lib files that reduce the TCB.   Are we assuming the linker will always optimize out un-called obj files out of the libc?

Reply:
Unresolving just to get an answer to my question above "Are we assuming the linker will always optimize out un-called obj files out of the libc?"   If answer is Yes, ok to re-resolve.

Reply:
Correct, that is the assumption. Thanks